api/src/main/java/org/cardanofoundation/rosetta/api/account/service/AccountServiceImpl.java

```diff
@@ -84,33 +98,60 @@ private AccountBalanceResponse findBalanceDataByAddressAndBlock(String address,
           if (CardanoAddressUtils.isStakeAddress(address)) {
             log.debug("Address is StakeAddress, get balance for {}", address);
-            List<StakeAddressBalance> balances = ledgerAccountService.findStakeAddressBalanceByAddressAndBlock(
-                address, blockDto.getNumber());
-            if (Objects.isNull(balances) || balances.isEmpty()) {
-              log.error("[findBalanceDataByAddressAndBlock] No balance found for {}", address);
-              throw ExceptionFactory.invalidAddressError();
-            }
-            return DataMapper.mapToStakeAddressBalanceResponse(blockDto, balances.getFirst());
+            BigInteger quantity = ledgerAccountService
+                .findStakeAddressBalanceQuantityByAddressAndBlock(address, blockDto.getNumber())
+                .orElse(BigInteger.ZERO);
+            return DataMapper.mapToStakeAddressBalanceResponse(blockDto, quantity);
           } else {
             log.debug("Address isn't StakeAddress");
-            List<AddressBalance> balances = ledgerAccountService.findBalanceByAddressAndBlock(
-                address, blockDto.getNumber());
+            List<AddressBalance> balances = ledgerAccountService
+                .findBalanceByAddressAndBlock(address, blockDto.getNumber());
             return DataMapper.mapToAccountBalanceResponse(blockDto, balances);
           }
         })
         .orElseThrow(ExceptionFactory::blockNotFoundException);
   }
 
-  private Optional<Block> findBlockOrLast(Long number, String hash) {
+  private Optional<BlockIdentifierExtended> findBlockOrLast(Long number, String hash) {
     if (number != null || hash != null) {
-      return ledgerBlockService.findBlock(number, hash);
+      return ledgerBlockService.findBlockIdentifier(number, hash);
+    } else {
+      return Optional.of(ledgerBlockService.findLatestBlockIdentifier());
+    }
+  }
+
+  private void validateCurrencies(List<Currency> currencies) {
+    for (Currency currency : currencies) {
+      String symbol = currency.getSymbol();
+      CurrencyMetadata metadata = currency.getMetadata();
+      if (!isTokenNameValid(symbol)) {
+        throw invalidTokenNameError("Given name is " + symbol);
+      }
+      if (!symbol.equals(Constants.ADA)
+          && (metadata == null || !isPolicyIdValid(String.valueOf(metadata.getPolicyId())))) {
+        String policyId = metadata == null ? null : metadata.getPolicyId();
+        throw invalidPolicyIdError("Given policy id is " + policyId);
+      }
+    }
+  }
+
+  private boolean isTokenNameValid(String name) {
+    return TOKEN_NAME_VALIDATION.matcher(name).matches() || isEmptyHexString(name);
+  }
+
+  private boolean isPolicyIdValid(String policyId) {
+    return POLICY_ID_VALIDATION.matcher(policyId).matches();
+  }
+
+  private List<Currency> filterRequestedCurrencies(List<Currency> currencies) {
+    if (currencies != null && currencies.stream().map(Currency::getSymbol)
```

Comment:
Also, can be refactored in functional style (without if(..))

Reply:
That's reasonable. Done.